nixos/doc/manual/development/option-types.section.md

```diff
@@ -279,6 +282,67 @@ Submodules are detailed in [Submodule](#section-option-types-submodule).
     more convenient and discoverable than expecting the module user to
     type-merge with the `attrsOf submodule` option.
 
+## Union types {#sec-option-types-unions}
+
+A union of types is a type such that a value is valid when it is valid for at least one of those types.
+
+If some values are instances of more than one of the types, it is not possible to distinguish which type they are meant to be instances of. If that's needed, consider using a [sum type](#sec-option-types-sums).
+
+`types.either` *`t1 t2`*
+
+:   Type *`t1`* or type *`t2`*, e.g. `with types; either int str`.
+    Multiple definitions cannot be merged.
+
+`types.oneOf` \[ *`t1 t2`* ... \]
+
+:   Type *`t1`* or type *`t2`* and so forth, e.g.
+    `with types; oneOf [ int str bool ]`. Multiple definitions cannot be
+    merged.
+
+`types.nullOr` *`t`*
+
+:   `null` or type *`t`*. Multiple definitions are merged according to
+    type *`t`*.
+
+
+## Sum types {#sec-option-types-sums}
+
+A sum type can be thought of, conceptually, as a *`types.enum`* where each valid item is paired with a type, through some value syntax.
+Nix does not have a built-in syntax for this pairing of a label and a type or value, so sum types may be represented in multiple ways.
+
+If the built-in Nix value types provide enough distinction, you simplify your syntax with a [union type](#sec-option-types-unions) instead.
+
+`types.attrTag` *`{ attr1 = t1; attr2 = t2; ... }`*
+
+:   An attribute set containing one attribute, whose name must be picked from
+    the attribute set (`attr1`, etc) and whose value must be of the accompanying
+    type.
+
+    This type appears in the documentation as _attribute-tagged union_.
+
+    Example:
+
+    ```nix
+    types.attrTag {
+      bounce = mkOption {
+        description = "Send back a packet explaining why it wasn't forwarded.";
+        type = submodule {
+          options.errorMessage = mkOption { … };
+        };
+      };
+      forward = mkOption {
+        description = "Forward the packet.";
+        type = submodule {
+          options.destination = mkOption { … };
+        };
+      };
+      ignore = mkOption {
+        description = "Drop the packet without sending anything back.";
+        type = submodule {};
+      };
+    }
+    ```
```

Comment:
An example of possible values for an option of this type would be helpful.

My best guess would be that it allows
```nix
{ bounce.errorMessage = "foo bar"; }
```
and
```
{ ignore = {}; }
```
?